### PR TITLE
Support for "attributes" in \yii\web\Request

### PR DESCRIPTION
**Pull request by rhertogh**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | -

This PR adds [PSR-7](https://www.php-fig.org/psr/psr-7/#:~:text=public%20function%20getAttributes()%3B) like support for "attributes" in `\yii\web\Request`.

### CODE REVIEW

File: framework/web/Request.php [hunk 1]
```diff
@@ -261,6 +261,13 @@ class Request extends \yii\base\Request
         'Front-End-Https' => ['on'], // Microsoft
     ];
 
+    /**
+     * @var mixed[] The request "attributes" can be used to for storage of any parameters derived from the request,
+     * e.g. the scopes of an Oauth request.
+     * @since 2.0.44
+     */
+    protected $_attributes = [];
```
Comment (WinterSilence) [inline framework/web/Request.php:269]: add `@property` tag to class

File: tests/framework/web/RequestTest.php [hunk 2]
```diff
@@ -1267,4 +1267,24 @@ public function testForwardedNotTrusted()
         $this->assertSame('10.0.0.1', $request->userIP, 'User IP fail!');
         $this->assertSame('http://yiiframework.com', $request->hostInfo, 'Host info fail!');
     }
+
```
Comment (WinterSilence) [inline tests/framework/web/RequestTest.php:1270]: add `@covers` tags to methid

**Comment by samdark** [comment]
What problem do you want to solve with attributes? Could it be solved currently by different means? Why attributes way of solving it is better?

**Comment by rhertogh** [comment]
> What problem do you want to solve with attributes? Could it be solved currently by different means? Why attributes way of solving it is better?

I'm building an Oauth2 extension. It extracts the Scopes from the JWT. The most safe and logical place to store them would be on the request. This kind of data belongs to the request and this way, if the request changes, also the scopes would be removed resulting in a much more secure solution.

**Comment by samdark** [comment]
@rhertogh what do you mean by "more secure solution"? Typically whole context is eliminated when PHP is used with Apache or php-fpm.

**Comment by rhertogh** [comment]
> what do you mean by "more secure solution"? Typically whole context is eliminated when PHP is used with Apache or php-fpm.

Normally yes, but if, for example, the it's being used with roadrunner or with other edge cases this might not be the case.

File: framework/web/Request.php (hunk 1 above)
Comment (rhertogh) [reply framework/web/Request.php:271]: @property-read tag added (since the attributes are not directly writable)

**Comment by samdark** [comment]
Do you use the app with RoadRunner?

**Comment by rhertogh** [comment]
> Do you use the app with RoadRunner?

No, but it will be a publicly available extension, so I can't predict the different environments and/or applications it will be used in.

**Comment by samdark** [comment]
Is that OAuth server?

**Comment by rhertogh** [comment]
> Is that OAuth server?

Yes

**Comment by samdark** [comment]
@yiisoft/reviewers need your opinions.

File: framework/web/Request.php (hunk 1 above)
Comment (WinterSilence) [reply framework/web/Request.php:271]: @rhertogh `@property-read array $attributes` and `@property mixed $attribute`

Comment (rhertogh) [reply framework/web/Request.php:271]: @WinterSilence How would `@property mixed $attribute` work? The `getAttribute($name, $default = null)` requires the `$name` parameter, is it still possible to define `attribute` as a property of the class?

Comment (WinterSilence) [reply framework/web/Request.php:271]: @rhertogh ups.. you right, all fine

**Comment by bizley** [comment]
I don't have a strong opinion on this. It's nice to have additional features but it comes with a cost of supporting them later on.

**Comment by My6UoT9** [comment]
It sounds to me like API access, for me it does not belong in \yii\web\request.
maybe \yii\api\request ? But I only looked superficially at it.

**Comment by rhertogh** [comment]
> It sounds to me like API access, for me it does not belong in \yii\web\request.
> maybe \yii\api\request ? But I only looked superficially at it.

The Oauth Scopes mentioned in this treat were an example. The PSR-7 standard provides more examples:

> ... e.g., the results of path match operations; the results of decrypting cookies; the results of deserializing non-form-encoded message bodies; etc. Attributes will be application and request specific ...

Therefore I would not regard the Request "attributes" an API specific feature.

**Comment by WinterSilence** [comment]
@rhertogh `yii\web\Request` not implements `Psr\Http\Message\Request` and already have properties for this

**Comment by rhertogh** [comment]
@WinterSilence 
> `yii\web\Request` not implements `Psr\Http\Message\Request` and already have properties for this

Do you mean the properties as defined in `\yii\base\Component` (e.g. `hasProperty()`)? In that case the "attributes" serve a different purpose. The attributes can been seen as metadata of the request (not a property thereof).

**Comment by WinterSilence** [comment]
@rhertogh I mean class already have properties for cookies, body data and etc. You should demonstrate why we need new property in `Request` instead using another "storages".

**Comment by rhertogh** [comment]
The cookies, body, etc. contain data that comes directly from the user and therefore should not be trusted. The attributes property can contain data dat is derived from, and directly belongs to, the request. For example (but not limited to) the scopes for Oauth. They are send in the request in the header via a JWT. A Yii extension could decode and validate the JWT and if the JWT is valid expose the decoded data via the attributes. The reason to store the data on the request is because it belongs to the request and, as mentioned earlier, if the request changes for any reason like for example when the app is running on roadrunner or for any other reason outside the scope of the extension, the request attributes should be removed too. Therefore the most logical place is to store them on the request itself.

**Comment by WinterSilence** [comment]
@rhertogh Why not use behavior? Then you can combine decoder and decoded header together.

**Comment by rhertogh** [comment]
@WinterSilence Yes, that's possible. The plan is to release the oauth2 server for both Yii version 2 and 3. This PR helps with future compatibility and allows extensions to be more easily ported to Yii3. As, if I understand it correctly, there are no behaviors in Yii3?

**Comment by samdark** [comment]
@rhertogh Yii3 is very different. There are request-response, middleware stack etc. etc. I'd not take it into account while considering this PR.

**Comment by WinterSilence** [comment]
@rhertogh name is only thing that unites yii2 and yii3. Talk with owners to add this futures as behavior - I'm think this possible to append.

**Comment by rhertogh** [comment]
> Yii3 is very different. There are request-response, middleware stack etc. etc. I'd not take it into account while considering this PR.

> name is only thing that unites yii2 and yii3. Talk with owners to add this futures as behavior - I'm think this more possible.

OK, it looks like a Yii3 version of the extension would need a rewrite anyway so then there is no need for the "attributes".

**Comment by WinterSilence** [comment]
@rhertogh  yii3 PSR-compatible - https://github.com/yiisoft/request-model/blob/master/src/RequestModel.php#L19

**Comment by samdark** [comment]
Yeah, that would be a bit too-much for Yii2, I think. Thanks for contributing anyway.

**Pull request closed without merging**